### PR TITLE
rp2040: Avoid screeches due to audio underflow during flash writes

### PR DESCRIPTION
**Pull request by jepler**
By pausing audio during flash writes, the worst screeching of #8121 is avoided. I don't consider this a full fix, but it greatly improves the by far most common scenario in which the problem occurs, and can't otherwise be worked around

Tested on rp2040 prop feather with a midi synth playing arpeggios. When writing to the flash e.g., with
```
dd bs=512 count=32 if=/dev/zero of=/media/jepler/CIRCUITPY/boop
```
the audio goes "tap tap tap tap" during the flash write instead of the squawking.

This isn't a 100% fix; it will still glitch out, including during USB enumeration which must be taking a long time without servicing background tasks. Add a delay if not usb-connected at startup ameliorates this greatly.

to work around the screech at startup, I have found that this is adequate:
```
def maybe_wait():
    deadline = time.monotonic() + 1
    while time.monotonic() < deadline and not usb_cdc.console.connected:
        pass
```

@todbot I know this problem has vexed you so your testing would be appreciated. It's still not perfect but I think you'll be able to judge if it's an improvement on the status quo.

### CODE REVIEW

**Comment by jepler** [comment]
fwiw I briefly tried implementing the "DMA controller sets its own next-DMA address" scott suggested but it didn't work right off the bat as a simple change.

It seems like either you'd end up using 4 DMA channels (2 to actually do transfers + 2 to set transfer start addresses), or you would have to use a different method than checking which DMA interrupt fired to know which memory buffer to fill.

File: ports/raspberrypi/audio_dma.h [hunk 1]
```diff
@@ -89,4 +89,7 @@ void audio_dma_pause(audio_dma_t *dma);
 void audio_dma_resume(audio_dma_t *dma);
 bool audio_dma_get_paused(audio_dma_t *dma);
 
+uint32_t audio_dma_pause_all(void);
+void audio_dma_unpause_cookie(uint32_t cookie);
```
Comment (dhalbert) [inline ports/raspberrypi/audio_dma.h:93]:
```suggestion
void audio_dma_unpause_channel_mask(uint32_t channel_mask);
```

File: ports/raspberrypi/supervisor/internal_flash.c [hunk 2]
```diff
@@ -97,9 +98,12 @@ void port_internal_flash_flush(void) {
     }
     // Make sure we don't have an interrupt while we do flash operations.
     common_hal_mcu_disable_interrupts();
+    // and audio DMA must be paused as well
+    uint32_t cookie = audio_dma_pause_all();
     flash_range_erase(CIRCUITPY_CIRCUITPY_DRIVE_START_ADDR + _cache_lba, SECTOR_SIZE);
     flash_range_program(CIRCUITPY_CIRCUITPY_DRIVE_START_ADDR + _cache_lba, _cache, SECTOR_SIZE);
     _cache_lba = NO_CACHE;
+    audio_dma_unpause_cookie(cookie);
```
Comment (dhalbert) [inline ports/raspberrypi/supervisor/internal_flash.c:106]:
```suggestion
    uint32_t channel_mask = audio_dma_pause_all();
    flash_range_erase(CIRCUITPY_CIRCUITPY_DRIVE_START_ADDR + _cache_lba, SECTOR_SIZE);
    flash_range_program(CIRCUITPY_CIRCUITPY_DRIVE_START_ADDR + _cache_lba, _cache, SECTOR_SIZE);
    _cache_lba = NO_CACHE;
    audio_dma_unpause_channel_mask(channel_mask);
```

**Review by dhalbert — CHANGES_REQUESTED** [review]
Code looks good to me. Suggestion of more descriptive name than `cookie`.

I would feel comfortable putting this in 8.2.0. How about you?

**Comment by dhalbert** [comment]
> I'd like for this to be in 8.2. Do I need to re-target the branch or is main still where 8.2 comes from?

No problem - it's still `main`. I will make a branch when I make 8.2.0 final.

**Review by dhalbert — APPROVED** [review]
Thanks! This is a big improvement!

**Pull request merged**